### PR TITLE
Argon2d4096 Mining Algorithm Change

### PR DESCRIPTION
**Pull request by nzsquirrell**
Introduces Argon2d mining, replacing Skein via versionbits (bit 6) consensus.
Uses the same Argon2d parameters as Unitus and Argentum.

Activation period on testnet is from 1st Jan 2019, mainnet is 15th February.

### CODE REVIEW

File: src/Makefile.am [hunk 1]
```diff
@@ -355,7 +319,56 @@ libbitcoin_consensus_a_SOURCES = \
...
+  crypto/argon2/best.c \
+  crypto/argon2/blake2.h \
+  crypto/argon2/blake2b.c \
+  crypto/argon2/blake2-impl.h \
+  crypto/argon2/core.c \
+  crypto/argon2/core.h \
+  crypto/argon2/encoding.c \
+  crypto/argon2/encoding.h \
+  crypto/argon2/thread.c \
+  crypto/argon2/thread.h
```
Comment (cryptapus) [inline src/Makefile.am]: I think we might need to include something like `ARGON_DIST` similar to `YESCRYPT_DIST` later in this file to make gitian and travis-ci happy.

**Comment by cryptapus** [comment]
Concept ACK, will test when able (nice to see you again).

File: src/Makefile.am (hunk 1 above)
Comment (nzsquirrell) [reply src/Makefile.am]: Fixed ... after multiple attempts!

**Comment by 8bitcoder** [comment]
Concept ACK.

**Comment by cryptapus** [comment]
ACK

To be more precise, partially tested ACK. Observed that `chainwork` is identical to master, `argon2d` appears to be present in RPC calls. I believe the next step would be to merge and activate on testnet if consensus is to proceed.

**Comment by cryptapus** [comment]
Additionally, I volunteer to backport this to 0.14 (if easily done) for anyone who needs it once merged. Although I think to keep github `Latest release` happy we would always need to tag the 0.14 version prior to the 0.16 version.

**Comment by 8bitcoder** [comment]
Visually reviewed changes. ACK for moving to testnet for more testing.

**Comment by nzsquirrell** [comment]
Given the issues faced by reaching consensus on this change (esp. some sha/scrypt pools that do not upgrade despite efforts to contact), this fork may need to be changed to a predetermined height hard-fork. Thoughts?

**Comment by cryptapus** [comment]
My personal preference would be for our current BIP9 consensus process (75%), or at least see where it puts us. If we find that we are slightly short we can maybe do something more aggressive like a UASF (or UAHF in this case)?

**Comment by roarde** [comment]
We need to modify BIP9 for this and other reasons. See #134. Can that be done properly before activation of this change is attempted?

**Comment by cryptapus** [comment]
@nzsquirrell If you're so inclined, this patch should enable a simple POW test for Argon2D:

    --- a/src/test/blockencodings_tests.cpp
    +++ b/src/test/blockencodings_tests.cpp
    @@ -92,6 +92,7 @@ BOOST_AUTO_TEST_CASE(SimpleCheckAlgo)
         BOOST_CHECK_EQUAL(block.GetPoWHash(ALGO_SKEIN,Params().GetConsensus()).ToString() , "c938e0667fbca99c5d57374c9c517e5257801257123b521c72c836632a585859");
         BOOST_CHECK_EQUAL(block.GetPoWHash(ALGO_QUBIT,Params().GetConsensus()).ToString() , "8f8e107327fca43c9c191ba0c8a326106cc65fe98596362b12455bf510632cc2");
         BOOST_CHECK_EQUAL(block.GetPoWHash(ALGO_YESCRYPT,Params().GetConsensus()).ToString() , "1d31a0941db6e1d2ab7503a7026dadd0dc01d49d12e561ca574c7fd945d23a7b");
    +    BOOST_CHECK_EQUAL(block.GetPoWHash(ALGO_ARGON2D,Params().GetConsensus()).ToString() , "956ee8ec3fb42aa627c7e70a1eeee88e399fcf5cec213c4c07aca1a51f0242a6");
     }
    
     BOOST_AUTO_TEST_CASE(SimpleRoundTripTest)

**Comment by cryptapus** [comment]
> We need to modify BIP9 for this and other reasons. See #134. Can that be done properly before activation of this change is attempted?

@roarde Your concerns are appreciated. I think it would be helpful to have some code to analyse such an improvement and discuss further. Perhaps as a separate MIP?

**Comment by cryptapus** [comment]
Is there any objection to merging this and #133 end of wk1905 to activate on testnet?

**Pull request merged**